contracts/CreditorProxy.sol

```diff
@@ -0,0 +1,409 @@
+     * in the entire system.
+     */
+    function setDebtKernel(address debtKernelAddress)
+        public
+        onlyOwner
+    {
+        debtKernel = DebtKernel(debtKernelAddress);
+    }
+
+
+    /*
+     * Submit debt order to DebtKernel if it is consensual with creditor's request
+     * Creditor signature in arguments is external creditor and not relevant to Debt Kernel
+     */
+    function fillCreditOrder(
+        address creditor,
+        address[6] orderAddresses, // repayment-router, debtor, uw, tc, p-token, relayer
+        uint[8] orderValues, // rr, salt, pa, uwFee, rFee, cFee, dFee, expTime
+        bytes32[1] orderBytes32, // tcParams
+        uint8[3] signaturesV, // debtV, credV, uwV
+        bytes32[3] signaturesR,
+        bytes32[3] signaturesS
+    )
+    public
+    whenNotPaused
+    returns (bytes32 _agreementId)
+    {
+        uint principalAmount = orderValues[2];
+        address principalToken = orderAddresses[4];
+
+        CreditorCommitment memory creditorCommitment = getCreditorCommitment(
+            creditor, orderAddresses, orderValues, orderBytes32
+        );
+
+        uint totalCreditorPayment = principalAmount.add(creditorCommitment.creditorFee);
+
+        if (!assertNoReplay(creditorCommitment.hash)) { 
+            LogError(uint8(Errors.CREDIT_ORDER_CANCELLED), creditor, creditorCommitment.hash);
+            return NULL_ISSUANCE_HASH; 
+        }
+
+        if (!isValidSignature(
+            creditor,
+            creditorCommitment.hash,
+            signaturesV[1],
+            signaturesR[1],
+            signaturesS[1]
+        )) {
+            LogError(uint8(Errors.CREDIT_ORDER_NON_CONSENSUAL), creditor, creditorCommitment.hash);
+            return NULL_ISSUANCE_HASH;
+        }
+
+        if (!assertExternalBalanceAndAllowanceInvariants(creditor, principalToken, totalCreditorPayment)) {
+            LogError(
+                uint8(Errors.CREDITOR_BALANCE_OR_ALLOWANCE_INSUFFICIENT),
+                creditor,
+                creditorCommitment.hash
+            );
+            return NULL_ISSUANCE_HASH;
+        }
+
+        // Transfer principal from creditor to creditorProxy
+        if (totalCreditorPayment > 0) {
+            require(transferTokensFrom(
+                principalToken,
+                creditor,
+                address(this),
+                totalCreditorPayment
+            ));
+        }
+
+        // Fill debt order with this contract playing the role of creditor
+        bytes32 agreementId = debtKernel.fillDebtOrder(
+            address(this),
+            orderAddresses,
+            orderValues,
+            orderBytes32,
+            signaturesV,
+            signaturesR,
+            signaturesS
+        );
+
+        // cancel credit order if fillDebtOrder succeeded
+        if (agreementId != NULL_ISSUANCE_HASH) {
+            creditOrderCancelled[creditorCommitment.hash] = true;
```

Comment:
can we create a separate mapping for successfully `filled` orders? I think right now we're keeping track of filled orders in a hacky way.  having a separate mapping for filled orders would also allow for better error messaging ("Order has already been filled" vs "Order has been canceled")

Reply:
Good thinking!